File: src/main/java/org/taktik/icure/dao/impl/FormTemplateDAOImpl.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.dao.impl;

import com.google.common.io.ByteStreams;
import org.apache.commons.codec.digest.DigestUtils;
import org.ektorp.AttachmentInputStream;
import org.ektorp.ComplexKey;
import org.ektorp.DocumentNotFoundException;
import org.ektorp.support.View;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.cache.CacheManager;
import org.springframework.stereotype.Repository;
import org.taktik.icure.dao.FormTemplateDAO;
import org.taktik.icure.dao.impl.ektorp.CouchDbICureConnector;
import org.taktik.icure.dao.impl.idgenerators.IDGenerator;
import org.taktik.icure.dao.impl.idgenerators.UUIDGenerator;
import org.taktik.icure.entities.FormTemplate;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.util.List;

/**
 * Created by aduchate on 02/02/13, 15:24
 */

@Repository("formTemplateDAO")
@View(name = "all", map = "function(doc) { if (doc.java_type == 'org.taktik.icure.entities.FormTemplate' && !doc.deleted) emit(doc._id, null )}")
class FormTemplateDAOImpl extends CachedDAOImpl<FormTemplate> implements FormTemplateDAO {
	private static final Logger log = LoggerFactory.getLogger(FormTemplateDAOImpl.class);

	private UUIDGenerator uuidGenerator;

	@Autowired
	public void setUuidGenerator(UUIDGenerator uuidGenerator) {
		this.uuidGenerator = uuidGenerator;
	}

	@Autowired
    public FormTemplateDAOImpl(@SuppressWarnings("SpringJavaAutowiringInspection") @Qualifier("couchdbBase") CouchDbICureConnector couchdb, IDGenerator idGenerator, @Qualifier("entitiesCacheManager") CacheManager cacheManager) {
        super(FormTemplate.class, couchdb, idGenerator, cacheManager);
        initStandardDesignDocument();
    }

    @Override
    @View(name = "by_userId_and_guid", map = "function(doc) { if (doc.java_type == 'org.taktik.icure.entities.FormTemplate' && !doc.deleted && doc.author) emit([doc.author,doc.guid], null )}")
    public List<FormTemplate> findByUserGuid(String userId, String guid, boolean loadLayout) {
        ComplexKey from = ComplexKey.of(userId, guid != null ? guid : "");
		ComplexKey to = ComplexKey.of(userId, guid != null ? guid : "\ufff0");
		List<FormTemplate> formTemplates = queryView("by_userId_and_guid", from, to);

		// invoke postLoad()
	    if (loadLayout) {
		    formTemplates.forEach(this::postLoad);
	    }

		return formTemplates;
	}

    @Override
    @View(name = "by_specialty_code_and_guid", map = "function(doc) { if (doc.java_type == 'org.taktik.icure.entities.FormTemplate' && !doc.deleted && doc.specialty) emit([doc.specialty.code,doc.guid], null )}")
    public List<FormTemplate> findBySpecialtyGuid(String specialityCode, String guid, boolean loadLayout) {

		List<FormTemplate> formTemplates;
		if (guid !=null) {
			ComplexKey key = ComplexKey.of(specialityCode, guid);
			formTemplates = queryView("by_specialty_code_and_guid", key);
		} else {
			ComplexKey from = ComplexKey.of(specialityCode, null);
			ComplexKey to = ComplexKey.of(specialityCode, ComplexKey.emptyObject());
			formTemplates = queryView("by_specialty_code_and_guid", from, to);
		}

		// invoke postLoad()
	    if (loadLayout) {
		    formTemplates.forEach(this::postLoad);
	    }

		return formTemplates;
	}


	public void evictFromCache(FormTemplate entity) {
		super.evictFromCache(entity);
	}

	public FormTemplate createFormTemplate(FormTemplate entity) {
		super.save(true, entity);
		return entity;
	}


	@Override
	protected void beforeSave(FormTemplate entity) {
		super.beforeSave(entity);

		if (entity.getLayout() != null) {
			String newLayoutAttachmentId = DigestUtils.sha256Hex(entity.getLayout());

			if (!newLayoutAttachmentId.equals(entity.getLayoutAttachmentId())) {
				entity.setLayoutAttachmentId(newLayoutAttachmentId);
				entity.setAttachmentDirty(true);
			}
		}
	}

	@Override
	protected void afterSave(FormTemplate entity) {
		super.afterSave(entity);

		if (entity.isAttachmentDirty() && entity.getLayout() != null && entity.getLayoutAttachmentId() != null) {
			AttachmentInputStream a = new AttachmentInputStream(entity.getLayoutAttachmentId(), new ByteArrayInputStream(entity.getLayout()), "application/json");
			entity.setRev(createAttachment(entity.getId(), entity.getRev(), a));
			entity.setAttachmentDirty(false);
		}
	}

	@Override
	public void postLoad(FormTemplate entity) {
		super.postLoad(entity);

		if (entity != null && entity.getLayoutAttachmentId() != null) {
			try {
				AttachmentInputStream attachmentIs = getAttachmentInputStream(entity.getId(), entity.getLayoutAttachmentId());
				byte[] layout = ByteStreams.toByteArray(attachmentIs);
				entity.setLayout(layout);
			} catch (IOException|DocumentNotFoundException e) {
				log.warn("Failed to obtain attachment(" + entity.getId() + ") for the doc id (" + entity.getLayoutAttachmentId() + ").");
			}
		}
	}
}

File: src/main/kotlin/org/taktik/icure/asyncdao/impl/FormTemplateDAOImpl.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */

package org.taktik.icure.asyncdao.impl

import java.io.IOException
import java.nio.ByteBuffer
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.map
import org.apache.commons.codec.digest.DigestUtils
import org.apache.commons.io.output.ByteArrayOutputStream
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Repository
import org.taktik.couchdb.annotation.View
import org.taktik.couchdb.entity.ComplexKey
import org.taktik.couchdb.id.IDGenerator
import org.taktik.couchdb.id.UUIDGenerator
import org.taktik.couchdb.queryViewIncludeDocsNoValue
import org.taktik.icure.asyncdao.FormTemplateDAO
import org.taktik.icure.entities.FormTemplate
import org.taktik.icure.properties.CouchDbProperties
import org.taktik.icure.spring.asynccache.AsyncCacheManager
import org.taktik.icure.utils.writeTo

/**
 * Created by aduchate on 02/02/13, 15:24
 */

@Repository("formTemplateDAO")
@View(name = "all", map = "function(doc) { if (doc.java_type == 'org.taktik.icure.entities.FormTemplate' && !doc.deleted) emit(null, doc._id)}")
internal class FormTemplateDAOImpl(
	couchDbProperties: CouchDbProperties,
	private val uuidGenerator: UUIDGenerator,
	@Qualifier("baseCouchDbDispatcher") couchDbDispatcher: CouchDbDispatcher,
	idGenerator: IDGenerator,
	@Qualifier("asyncCacheManager") asyncCacheManager: AsyncCacheManager
) : GenericDAOImpl<FormTemplate>(couchDbProperties, FormTemplate::class.java, couchDbDispatcher, idGenerator), FormTemplateDAO {

	@View(name = "by_userId_and_guid", map = "function(doc) { if (doc.java_type == 'org.taktik.icure.entities.FormTemplate' && !doc.deleted && doc.author) emit([doc.author,doc.guid], null )}")
	override fun listFormTemplatesByUserGuid(userId: String, guid: String?, loadLayout: Boolean): Flow<FormTemplate> = flow {
		val client = couchDbDispatcher.getClient(dbInstanceUrl)

		val from = ComplexKey.of(userId, guid ?: "")
		val to = ComplexKey.of(userId, guid ?: "\ufff0")
		val formTemplates = client.queryViewIncludeDocsNoValue<Array<String>, FormTemplate>(createQuery(client, "by_userId_and_guid").startKey(from).endKey(to).includeDocs(true)).map { it.doc }

		// invoke postLoad()
		emitAll(
			if (loadLayout) {
				formTemplates.map {
					this@FormTemplateDAOImpl.postLoad(it)
				}
			} else formTemplates
		)
	}

	@View(name = "by_guid", map = "function(doc) { if (doc.java_type == 'org.taktik.icure.entities.FormTemplate' && !doc.deleted) emit(doc.guid, null )}")
	override fun listFormsByGuid(guid: String, loadLayout: Boolean): Flow<FormTemplate> = flow {
		val client = couchDbDispatcher.getClient(dbInstanceUrl)

		val formTemplates = client.queryViewIncludeDocsNoValue<String, FormTemplate>(createQuery(client, "by_guid").key(guid).includeDocs(true)).map { it.doc }

		emitAll(
			if (loadLayout) {
				formTemplates.map {
					this@FormTemplateDAOImpl.postLoad(it)
				}
			} else formTemplates
		)
	}

	@View(name = "by_specialty_code_and_guid", map = "function(doc) { if (doc.java_type == 'org.taktik.icure.entities.FormTemplate' && !doc.deleted && doc.specialty) emit([doc.specialty.code,doc.guid], null )}")
	override fun listFormsBySpecialtyAndGuid(specialityCode: String, guid: String?, loadLayout: Boolean): Flow<FormTemplate> = flow {
		val client = couchDbDispatcher.getClient(dbInstanceUrl)

		val formTemplates = if (guid != null) {
			val key = ComplexKey.of(specialityCode, guid)
			client.queryViewIncludeDocsNoValue<Array<String>, FormTemplate>(createQuery(client, "by_specialty_code_and_guid").key(key).includeDocs(true)).map { it.doc }
		} else {
			val from = ComplexKey.of(specialityCode, null)
			val to = ComplexKey.of(specialityCode, ComplexKey.emptyObject())
			client.queryViewIncludeDocsNoValue<Array<String>, FormTemplate>(createQuery(client, "by_specialty_code_and_guid").startKey(from).endKey(to).includeDocs(true)).map { it.doc }
		}

		emitAll(
			if (loadLayout) {
				formTemplates.map {
					this@FormTemplateDAOImpl.postLoad(it)
				}
			} else formTemplates
		)
	}

	override suspend fun createFormTemplate(entity: FormTemplate): FormTemplate {
		super.save(true, entity)
		return entity
	}

	override suspend fun beforeSave(entity: FormTemplate) =
		super.beforeSave(entity).let { formTemplate ->
			if (formTemplate.layout != null) {
				val newAttachmentId = DigestUtils.sha256Hex(formTemplate.layout)

				if (newAttachmentId != formTemplate.layoutAttachmentId && formTemplate.rev != null && formTemplate.layoutAttachmentId != null) {
					formTemplate.attachments?.containsKey(formTemplate.layoutAttachmentId)?.takeIf { it }?.let {
						formTemplate.copy(
							rev = deleteAttachment(formTemplate.id, formTemplate.rev, formTemplate.layoutAttachmentId),
							attachments = formTemplate.attachments - formTemplate.layoutAttachmentId,
							layoutAttachmentId = newAttachmentId,
							isAttachmentDirty = true
						)
					} ?: formTemplate.copy(
						layoutAttachmentId = newAttachmentId,
						isAttachmentDirty = true
					)
				} else
					formTemplate
			} else {
				if (formTemplate.layoutAttachmentId != null && formTemplate.rev != null) {
					formTemplate.copy(
						rev = deleteAttachment(formTemplate.id, formTemplate.rev, formTemplate.layoutAttachmentId),
						layoutAttachmentId = null,
						isAttachmentDirty = false
					)
				} else formTemplate
			}
		}

	override suspend fun afterSave(entity: FormTemplate) =
		super.afterSave(entity).let { formTemplate ->
			if (formTemplate.isAttachmentDirty && formTemplate.layoutAttachmentId != null && formTemplate.rev != null && formTemplate.layout != null) {
				createAttachment(formTemplate.id, formTemplate.layoutAttachmentId, formTemplate.rev, "application/json", flowOf(ByteBuffer.wrap(formTemplate.layout))).let {
					formTemplate.copy(
						rev = it,
						isAttachmentDirty = false
					)
				}
			} else formTemplate
		}

	override suspend fun postLoad(entity: FormTemplate) =
		super.postLoad(entity).let { formTemplate ->
			val formTemplateLayout = formTemplate.templateLayoutAttachmentId?.let {laId ->
				val attachmentFlow = getAttachment(formTemplate.id, laId, formTemplate.rev)
				ByteArrayOutputStream().use {
					attachmentFlow.writeTo(it)
					it.toByteArray()
				}
			}

			val formLayout = formTemplate.layoutAttachmentId?.takeIf { formTemplateLayout == null }?.let {laId ->
				val attachmentFlow = getAttachment(formTemplate.id, laId, formTemplate.rev)
				ByteArrayOutputStream().use {
					attachmentFlow.writeTo(it)
					it.toByteArray()
				}
			}

			if (formTemplateLayout != null || formLayout != null) {
				try {
					formTemplate.copy(templateLayout = formTemplateLayout, layout = formLayout)
				} catch (e: IOException) {
					formTemplate //Could not load
				}
			} else formTemplate
		}

	companion object {
		val log: Logger = LoggerFactory.getLogger(FormTemplateDAOImpl::class.java)
	}
}
